### ISSUE TITLE
perf: nexmark q17

### ISSUE THREAD

**Issue by lmatz** [risingwavelabs/risingwave#7351]
query:
```
CREATE MATERIALIZED VIEW nexmark_q17 AS
SELECT
    auction,
    to_char(date_time, 'YYYY-MM-DD') AS day,
    count(*) AS total_bids,
    count(*) filter (where price < 10000) AS rank1_bids,
    count(*) filter (where price >= 10000 and price < 1000000) AS rank2_bids,
    count(*) filter (where price >= 1000000) AS rank3_bids,
    min(price) AS min_price,
    max(price) AS max_price,
    avg(price) AS avg_price,
    sum(price) AS sum_price
FROM bid
GROUP BY auction, to_char(date_time, 'YYYY-MM-DD');
```

plan:
```
 StreamMaterialize { columns: [auction, day, total_bids, rank1_bids, rank2_bids, rank3_bids, min_price, max_price, avg_price, sum_price], pk_columns: [auction, day] }
 └─StreamProject { exprs: [Field(bid, 0:Int32), ToChar(Field(bid, 5:Int32), 'YYYY-MM-DD':Varchar), count, count filter((Field(bid, 2:Int32) < 10000:Int32)), count filter((Field(bid, 2:Int32) >= 10000:Int32) AND (Field(bid, 2:Int32) < 1000000:Int32)), count filter((Field(bid, 2:Int32) >= 1000000:Int32)), min(Field(bid, 2:Int32)), max(Field(bid, 2:Int32)), (sum(Field(bid, 2:Int32)) / count(Field(bid, 2:Int32))), sum(Field(bid, 2:Int32))] }
   └─StreamAppendOnlyHashAgg { group_key: [Field(bid, 0:Int32), ToChar(Field(bid, 5:Int32), 'YYYY-MM-DD':Varchar)], aggs: [count, count, count filter((Field(bid, 2:Int32) < 10000:Int32)), count filter((Field(bid, 2:Int32) >= 10000:Int32) AND (Field(bid, 2:Int32) < 1000000:Int32)), count filter((Field(bid, 2:Int32) >= 1000000:Int32)), min(Field(bid, 2:Int32)), max(Field(bid, 2:Int32)), sum(Field(bid, 2:Int32)), count(Field(bid, 2:Int32)), sum(Field(bid, 2:Int32))] }
     └─StreamExchange { dist: HashShard(Field(bid, 0:Int32), ToChar(Field(bid, 5:Int32), 'YYYY-MM-DD':Varchar)) }
       └─StreamProject { exprs: [Field(bid, 0:Int32), ToChar(Field(bid, 5:Int32), 'YYYY-MM-DD':Varchar), Field(bid, 2:Int32), _row_id] }
         └─StreamFilter { predicate: (event_type = 2:Int32) }
           └─StreamRowIdGen { row_id_index: 4 }
             └─StreamSource { source: "nexmark", columns: ["event_type", "person", "auction", "bid", "_row_id"] }
```

**Comment by lmatz** [comment]
Notice that `to_char` is optimzed before https://github.com/risingwavelabs/risingwave/pull/7056
But it may still take the most time, compared to other aggregation functions, 
see results of expression benchmark at https://github.com/risingwavelabs/risingwave/issues/6868

**Comment by fuyufjh** [comment]
Do these aggregate functions share one state table? Ideally, this could reduce IOPS to 1/8. @st1page @richardchien 

```
    count(*) AS total_bids,
    count(*) filter (where price < 10000) AS rank1_bids,
    count(*) filter (where price >= 10000 and price < 1000000) AS rank2_bids,
    count(*) filter (where price >= 1000000) AS rank3_bids,
    min(price) AS min_price,
    max(price) AS max_price,
    avg(price) AS avg_price,
    sum(price) AS sum_price
```

**Comment by fuyufjh** [comment]
Yes. 

```
dev=> show internal tables;
                         Name
------------------------------------------------------
 __internal_nexmark_q17_2002_sourceinternaltable_3007
 __internal_nexmark_q17_2001_hashaggresult_3006
```

**Comment by richardchien** [comment]
Yes. Since their states are all `ValueState`, they share one result table.

**Comment by st1page** [comment]
> Yes. Since their states are all `ValueState`, they share one result table.

FYI, the stream plan:
```
    Fragment 0
      StreamMaterialize { columns: [auction, day, total_bids, rank1_bids, rank2_bids, rank3_bids, min_price, max_price, avg_price, sum_price], pk_columns: [auction, day] }
          materialized table: 4294967294
        StreamProject { exprs: [bid.auction, ToChar(bid.date_time, 'YYYY-MM-DD':Varchar), count, count filter((bid.price < 10000:Int32)), count filter((bid.price >= 10000:Int32) AND (bid.price < 1000000:Int32)), count filter((bid.price >= 1000000:Int32)), min(bid.price), max(bid.price), (sum(bid.price) / count(bid.price)), sum(bid.price)] }
          StreamAppendOnlyHashAgg { group_key: [bid.auction, ToChar(bid.date_time, 'YYYY-MM-DD':Varchar)], aggs: [count, count, count filter((bid.price < 10000:Int32)), count filter((bid.price >= 10000:Int32) AND (bid.price < 1000000:Int32)), count filter((bid.price >= 1000000:Int32)), min(bid.price), max(bid.price), sum(bid.price), count(bid.price), sum(bid.price)] }
              result table: 0, state tables: []
            StreamExchange Hash([0, 1]) from 1

    Fragment 1
      StreamProject { exprs: [bid.auction, ToChar(bid.date_time, 'YYYY-MM-DD':Varchar), bid.price, bid._row_id] }
        Chain { table: bid, columns: [bid.auction, bid.price, bid.date_time, bid._row_id], pk: [bid._row_id], dist: UpstreamHashShard(bid._row_id) }
          Upstream
          BatchPlanNode

     Table 0 { columns: [bid.auction, ToChar(bid.date_time, 'YYYY-MM-DD':Varchar), count, count_0, count filter((bid.price < 10000:Int32)), count filter((bid.price >= 10000:Int32) AND (bid.price < 1000000:Int32)), count filter((bid.price >= 1000000:Int32)), min(bid.price), max(bid.price), sum(bid.price), count(bid.price), sum(bid.price)_0], primary key: [$0 ASC, $1 ASC], value indices: [2, 3, 4, 5, 6, 7, 8, 9, 10, 11], distribution key: [0, 1] }
     Table 4294967294 { columns: [auction, day, total_bids, rank1_bids, rank2_bids, rank3_bids, min_price, max_price, avg_price, sum_price], primary key: [$0 ASC, $1 ASC], value indices: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9], distribution key: [0, 1] }

```

**Comment by kwannoel** [comment]
Flink Query Plans:
```
== Optimized Physical Plan ==
Sink(table=[default_catalog.default_database.discard_sink], fields=[auction, day, total_bids, rank1_bids, rank2_bids, rank3_bids, min_price, max_price, avg_price, sum_price])
+- GroupAggregate(groupBy=[auction, day], select=[auction, day, COUNT(*) AS total_bids, COUNT(*) FILTER $f2 AS rank1_bids, COUNT(*) FILTER $f3 AS rank2_bids, COUNT(*) FILTER $f4 AS rank3_bids, MIN(price) AS min_price, MAX(price) AS max_price, AVG(price) AS avg_price, SUM(price) AS sum_price])
   +- Exchange(distribution=[hash[auction, day]])
      +- Calc(select=[bid.auction AS auction, DATE_FORMAT(CAST(dateTime AS TIMESTAMP(3)), _UTF-16LE'yyyy-MM-dd') AS day, IS TRUE(<(bid.price, 10000)) AS $f2, IS TRUE(SEARCH(bid.price, Sarg[[10000..1000000)])) AS $f3, IS TRUE(>=(bid.price, 1000000)) AS $f4, bid.price AS price], where=[=(event_type, 2)])
         +- WatermarkAssigner(rowtime=[dateTime], watermark=[-(dateTime, 4000:INTERVAL SECOND)])
            +- Calc(select=[event_type, bid, CASE(=(event_type, 0), person.dateTime, =(event_type, 1), auction.dateTime, bid.dateTime) AS dateTime])
               +- TableSourceScan(table=[[default_catalog, default_database, datagen]], fields=[event_type, person, auction, bid])

== Optimized Execution Plan ==
Sink(table=[default_catalog.default_database.discard_sink], fields=[auction, day, total_bids, rank1_bids, rank2_bids, rank3_bids, min_price, max_price, avg_price, sum_price])
+- GroupAggregate(groupBy=[auction, day], select=[auction, day, COUNT(*) AS total_bids, COUNT(*) FILTER $f2 AS rank1_bids, COUNT(*) FILTER $f3 AS rank2_bids, COUNT(*) FILTER $f4 AS rank3_bids, MIN(price) AS min_price, MAX(price) AS max_price, AVG(price) AS avg_price, SUM(price) AS sum_price])
   +- Exchange(distribution=[hash[auction, day]])
      +- Calc(select=[bid.auction AS auction, DATE_FORMAT(CAST(dateTime AS TIMESTAMP(3)), 'yyyy-MM-dd') AS day, (bid.price < 10000) IS TRUE AS $f2, SEARCH(bid.price, Sarg[[10000..1000000)]) IS TRUE AS $f3, (bid.price >= 1000000) IS TRUE AS $f4, bid.price AS price], where=[(event_type = 2)])
         +- WatermarkAssigner(rowtime=[dateTime], watermark=[(dateTime - 4000:INTERVAL SECOND)])
            +- Calc(select=[event_type, bid, CASE((event_type = 0), person.dateTime, (event_type = 1), auction.dateTime, bid.dateTime) AS dateTime])
               +- TableSourceScan(table=[[default_catalog, default_database, datagen]], fields=[event_type, person, auction, bid])
```

**Comment by kwannoel** [comment]
Updates for this week:
- Did some benchmarking to reproduce poor performance.
  - Not sure if my RW environment is optimal...
  - Flink: 463.56k vs RW: 300k
  - Instance: C5.4x large.
- Try again next week with the recommended instance (C5.2), see if I can reproduce the benchmark results.
- Took a look at `to_char` seems like the bulk of time is spent in this code block:
   https://github.com/risingwavelabs/risingwave/blob/849ffb803a1977db1b54534ada1b256ed9d4b392/src/expr/src/vector_op/to_char.rs#L62-L65

   Commenting it out the time decreases from `250->11μ`. Wonder if there's some way to optimize it.

---

Our to char expression just gets the day: `to_char(date_time, 'YYYY-MM-DD') as day`.
Currently I think our data won't be distributed by `day`, since nexmark runs a few minutes / query.

Will omitting date_time as group key have any significant performance impact?

Seems like it will eventually be done: https://github.com/risingwavelabs/risingwave/issues/3255

**Comment by fuyufjh** [comment]
Can we do a profile to get a flamegraph?

**Comment by kwannoel** [comment]
c5.2x large (8vCPU 16gb ram)

Throughput: 147315 r/s

[cpu flamegraph](https://user-images.githubusercontent.com/47273164/215680259-3e1d425d-4274-4f89-bc36-8d31d051662e.svg)

Memory flamegraph (upload soon, in progress)

![Screenshot 2023-01-31 at 2 12 14 PM](https://user-images.githubusercontent.com/47273164/215681516-04f7e0e4-cfbe-490f-9c3e-26fc4ead24cc.png)
![Screenshot 2023-01-31 at 2 12 20 PM](https://user-images.githubusercontent.com/47273164/215681523-724dce7e-9787-400d-8e61-cef34ba06574.png)
![Screenshot 2023-01-31 at 2 12 31 PM](https://user-images.githubusercontent.com/47273164/215681531-875f7810-d203-42ba-a639-a56ae94bdc77.png)

**Comment by fuyufjh** [comment]
Thanks! So the performance bottleneck is `to_char` right? Sorry the framegraph screenshot seems not to cover that part.

**Comment by kwannoel** [comment]
> Thanks! So the performance bottleneck is `to_char` right? Sorry the flamegraph screenshot seems not to cover that part.

From the flamegraph, I can't see any invocation of `to_char` / `ToChar`. Is there some way to get its symbols included?

In the screenshot, if we zoom in to the top right area, there are two areas with high cost:
1. `collect_vec`
2. `ValueState::apply_chunk`

![Screenshot 2023-01-31 at 5 50 12 PM](https://user-images.githubusercontent.com/47273164/215726959-1ffd3827-19de-4f8d-a7e8-5845d321f4f4.png)

Zooming in on `ValueState::apply_chunk`:
![Screenshot 2023-01-31 at 6 01 41 PM](https://user-images.githubusercontent.com/47273164/215729105-9ad1f1b0-fd2a-4520-8423-3a3b017ce5d4.png)

Cost is mainly from `StreamingFoldAgg::apply_batch`.

For that specific task which takes up 9% cpu time, we calculate the components:
```
StreamingFoldAgg<I64Array, I64Array, S>
time = 1.19 / 9 = 13%

StreamingFoldAgg<Decimal, I64Array, S> 
time = 0.78 / 9 = 9%

StreamingRowCount 
time = 0.58 / 9 = 6%
```

So does not seem like `to_char` is bottleneck actually? I will still investigate by running `to_char` alone and see.

**Comment by lmatz** [comment]
oh, the plan may be different,
RW `count` four times while Flink probably only `count` once 🤔 
correct me if I am wrong

**Comment by st1page** [comment]
I think the CPU is not the bottleneck here. The CPU usage is just 500% and the machine has 8 cores. Maybe we should focus on why the CPU is not running at full load and how it affects the e2e throughput (stateStore's IO or communication)

**Comment by KeXiangWang** [comment]
I cannot find parallelism info in the experiements. Since the source actors' number is 4(as the attached figure indicated), I guess the parallelism is 4, so there are 4 agg operators. Normally each actor takes at most 100% CPU usage. The source actor’s CPU load is low, so it seems reasonable to have 500% CPU usage.

A simple way to verify this is to increase the parallelism and then check the new CPU usage. Run this command in psql client before create MV to set parallelism:
```
set streaming_parallelism=8;
```

<img width="408" alt="image" src="https://user-images.githubusercontent.com/32863625/217156780-73339f38-a6bb-4b05-86e0-ed4413e2c2c9.png">

**Comment by lmatz** [comment]
> I cannot find parallelism info in the experiements. Since the source actors' number is 4(as the attached figure indicated), I guess the parallelism is 4, so there are 4 agg operators. Normally each actor takes at most 100% CPU usage. The source actor’s CPU load is low, so it seems reasonable to have 500% CPU usage.


Also, it seems only two out of four source actors are reading something

Together the fact
![SCR-20230207-j07](https://user-images.githubusercontent.com/5791930/217158342-08be4d47-63e5-483f-8a34-8e17ff9ef427.png)

Does it seem the number of Kafka partitions maybe 2?
If true, we may also adjust the number of Kafka partitions together with `set streaming_parallelism=8;`

**Comment by kwannoel** [comment]
Ah yes seems it was bottlenecked by source, now CPU can hit `800%`:

<img width="1456" alt="Screenshot 2023-02-07 at 2 22 24 PM" src="https://user-images.githubusercontent.com/47273164/217164619-06cf478c-83ad-4283-86a3-0a68410a976c.png">

But throughput is not ideal, @huangjw806 managed to get 224K r/s previously:

<img width="868" alt="Screenshot 2023-02-07 at 2 23 44 PM" src="https://user-images.githubusercontent.com/47273164/217164840-4323fcc4-61d4-4db0-b4bf-806e14ea0c25.png">

**Comment by KeXiangWang** [comment]
Are you using internal nexmark(not using nexmark-bench)? If yes, then there will be a lot of CPU time spent on generating data. @huangjw806 's work is based on nexmark-bench, where's the source's load is pretty low. However, it's a little bit weird that in Jianwei's experiments, the CPU usage is below 600%. It should be able to reach 800%.

**Comment by kwannoel** [comment]
> Are you using internal nexmark(not using nexmark-bench)? If yes, then there will be a lot of CPU time spent on generating data. @huangjw806 's work is based on nexmark-bench, where's the source's load is pretty low. 

Yes, Im using nexmark bench.

> However, it's a little bit weird that in Jianwei's experiments, the CPU usage is below 600%. It should be able to reach 800%. 

I think that's because parallelism=4 (default) in his experiments.

**Comment by kwannoel** [comment]
> Flink Query Plans:
> 
> ```
> == Optimized Physical Plan ==
> Sink(table=[default_catalog.default_database.discard_sink], fields=[auction, day, total_bids, rank1_bids, rank2_bids, rank3_bids, min_price, max_price, avg_price, sum_price])
> +- GroupAggregate(groupBy=[auction, day], select=[auction, day, COUNT(*) AS total_bids, COUNT(*) FILTER $f2 AS rank1_bids, COUNT(*) FILTER $f3 AS rank2_bids, COUNT(*) FILTER $f4 AS rank3_bids, MIN(price) AS min_price, MAX(price) AS max_price, AVG(price) AS avg_price, SUM(price) AS sum_price])
>    +- Exchange(distribution=[hash[auction, day]])
>       +- Calc(select=[bid.auction AS auction, DATE_FORMAT(CAST(dateTime AS TIMESTAMP(3)), _UTF-16LE'yyyy-MM-dd') AS day, IS TRUE(<(bid.price, 10000)) AS $f2, IS TRUE(SEARCH(bid.price, Sarg[[10000..1000000)])) AS $f3, IS TRUE(>=(bid.price, 1000000)) AS $f4, bid.price AS price], where=[=(event_type, 2)])
>          +- WatermarkAssigner(rowtime=[dateTime], watermark=[-(dateTime, 4000:INTERVAL SECOND)])
>             +- Calc(select=[event_type, bid, CASE(=(event_type, 0), person.dateTime, =(event_type, 1), auction.dateTime, bid.dateTime) AS dateTime])
>                +- TableSourceScan(table=[[default_catalog, default_database, datagen]], fields=[event_type, person, auction, bid])
> 
> == Optimized Execution Plan ==
> Sink(table=[default_catalog.default_database.discard_sink], fields=[auction, day, total_bids, rank1_bids, rank2_bids, rank3_bids, min_price, max_price, avg_price, sum_price])
> +- GroupAggregate(groupBy=[auction, day], select=[auction, day, COUNT(*) AS total_bids, COUNT(*) FILTER $f2 AS rank1_bids, COUNT(*) FILTER $f3 AS rank2_bids, COUNT(*) FILTER $f4 AS rank3_bids, MIN(price) AS min_price, MAX(price) AS max_price, AVG(price) AS avg_price, SUM(price) AS sum_price])
>    +- Exchange(distribution=[hash[auction, day]])
>       +- Calc(select=[bid.auction AS auction, DATE_FORMAT(CAST(dateTime AS TIMESTAMP(3)), 'yyyy-MM-dd') AS day, (bid.price < 10000) IS TRUE AS $f2, SEARCH(bid.price, Sarg[[10000..1000000)]) IS TRUE AS $f3, (bid.price >= 1000000) IS TRUE AS $f4, bid.price AS price], where=[(event_type = 2)])
>          +- WatermarkAssigner(rowtime=[dateTime], watermark=[(dateTime - 4000:INTERVAL SECOND)])
>             +- Calc(select=[event_type, bid, CASE((event_type = 0), person.dateTime, (event_type = 1), auction.dateTime, bid.dateTime) AS dateTime])
>                +- TableSourceScan(table=[[default_catalog, default_database, datagen]], fields=[event_type, person, auction, bid])
> ```

In the flink plan I see that we merge `auction, day` into `GroupAggregate`. Whereas for RW query plan, we have an extra `StreamProject`. Wonder if that has significant overhead.

**Comment by lmatz** [comment]
> In the flink plan I see that we merge auction, day into GroupAggregate. Whereas for RW query plan, we have an extra StreamProject. Wonder if that has significant overhead.

Hard to say, it seems RW uses this extra project to calculate AVG, i.e. SUM/COUNT, seems to be the only purpose.

**Comment by st1page** [comment]
@kwannoel do you mean the re-compute of the `TO_CHAR`? I think we can avoid it, but if we can not see it in the flame graph, it can not be the significant isuuse

**Comment by kwannoel** [comment]
> > In the flink plan I see that we merge auction, day into GroupAggregate. Whereas for RW query plan, we have an extra StreamProject. Wonder if that has significant overhead.
> 
> Hard to say, it seems RW uses this extra project to calculate AVG, i.e. SUM/COUNT, seems to be the only purpose.

I'm wondering if we reuse the `sum/count` states. 🤔 Will take a look.

**Comment by lmatz** [comment]
But wonder if `TO_CHAR` is able to appear in the flamegraph though due to `Box<dyn trait>`? Suppose it really consumes a significant amount of time

**Comment by kwannoel** [comment]
> But wonder if `TO_CHAR` is able to appear in the flamegraph though due to `Box<dyn trait>`? Suppose it really consumes a significant amount of time

I shall just remove it, and see the performance 🤣

**Comment by richardchien** [comment]
> I'm wondering if we reuse the `sum/count` states. 🤔 Will take a look, but seems like no from the rw plan:
> 
> ```
> sum(Field(bid, 2:Int32)), count(Field(bid, 2:Int32)), sum(Field(bid, 2:Int32))
> ```
> 
> cc @richardchien

I think we are not reusing the `sum` when rewriting `avg` to `sum / count`.

**Comment by kwannoel** [comment]
> > But wonder if `TO_CHAR` is able to appear in the flamegraph though due to `Box<dyn trait>`? Suppose it really consumes a significant amount of time
> 
> I shall just remove it, and see the performance 🤣

Made up query:
```sql
CREATE SINK nexmark_q17 AS
    SELECT auction,
           date_time IS NOT NULL                            AS day,
           count(*)                                                   AS total_bids,
           count(*) filter (where price < 10000)                      AS rank1_bids,
           count(*) filter (where price >= 10000 and price < 1000000) AS rank2_bids,
           count(*) filter (where price >= 1000000)                   AS rank3_bids,
           min(price)                                                 AS min_price,
           max(price)                                                 AS max_price,
           avg(price)                                                 AS avg_price,
           sum(price)                                                 AS sum_price
    FROM bid
    GROUP BY auction, date_time IS NOT NULL
    WITH ( connector = 'blackhole' );
```

Seems like performance becomes much better (see the third hill, previous 2 is other bench):
![Screenshot 2023-02-14 at 6 15 22 PM](https://user-images.githubusercontent.com/47273164/218706222-be5c7c81-6489-426d-9f17-a513a1ecce9d.png)

**Comment by kwannoel** [comment]
I guess we should optimize 2 things in order of priority:
1. Optimize `to_char`
2. Reuse `sum`.

**Comment by BugenZhao** [comment]
> But wonder if `TO_CHAR` is able to appear in the flamegraph though due to `Box<dyn trait>`?

I guess so. Note that all executor streams are `Box<dyn Stream>` as well. 🤔

**Comment by BugenZhao** [comment]
I'm quite curious why there must be a `to_char` in the group key. It seems all the query wants to do is truncate the timestamp into a date. Will we gain better performance if we rewrite it with `date_time::date`? It should even have the same representation in the result. 😄

**Comment by lmatz** [comment]
> rewrite it with date_time::date

In reality, we should if it is better
For the sake of benchmark comparison, we try to do exactly the same as Flink does.

**Comment by st1page** [comment]
> @kwannoel do you mean the re-compute of the `TO_CHAR`? I think we can avoid it, but if we can not see it in the flame graph, it can not be the significant issue

Sorry for my wrong information and misleading. In the plan, only the `StreamProject` before `StreamAppendOnlyHashAgg` does the evaluation of the `TO_CHAR` and the `StreamProject` before `StreamMaterialize` is just get the column from Agg's output. The `ToChar(Field(bid, 5:Int32), 'YYYY-MM-DD':Varchar)` is just the column's name!
```
 StreamMaterialize { columns: [auction, day, total_bids, rank1_bids, rank2_bids, rank3_bids, min_price, max_price, avg_price, sum_price], pk_columns: [auction, day] }
 └─StreamProject { exprs: [Field(bid, 0:Int32), ToChar(Field(bid, 5:Int32), 'YYYY-MM-DD':Varchar), count, count filter((Field(bid, 2:Int32) < 10000:Int32)), count filter((Field(bid, 2:Int32) >= 10000:Int32) AND (Field(bid, 2:Int32) < 1000000:Int32)), count filter((Field(bid, 2:Int32) >= 1000000:Int32)), min(Field(bid, 2:Int32)), max(Field(bid, 2:Int32)), (sum(Field(bid, 2:Int32)) / count(Field(bid, 2:Int32))), sum(Field(bid, 2:Int32))] }
   └─StreamAppendOnlyHashAgg { group_key: [Field(bid, 0:Int32), ToChar(Field(bid, 5:Int32), 'YYYY-MM-DD':Varchar)], aggs: [count, count, count filter((Field(bid, 2:Int32) < 10000:Int32)), count filter((Field(bid, 2:Int32) >= 10000:Int32) AND (Field(bid, 2:Int32) < 1000000:Int32)), count filter((Field(bid, 2:Int32) >= 1000000:Int32)), min(Field(bid, 2:Int32)), max(Field(bid, 2:Int32)), sum(Field(bid, 2:Int32)), count(Field(bid, 2:Int32)), sum(Field(bid, 2:Int32))] }
     └─StreamExchange { dist: HashShard(Field(bid, 0:Int32), ToChar(Field(bid, 5:Int32), 'YYYY-MM-DD':Varchar)) }
       └─StreamProject { exprs: [Field(bid, 0:Int32), ToChar(Field(bid, 5:Int32), 'YYYY-MM-DD':Varchar), Field(bid, 2:Int32), _row_id] }
         └─StreamFilter { predicate: (event_type = 2:Int32) }
           └─StreamRowIdGen { row_id_index: 4 }
             └─StreamSource { source: "nexmark", columns: ["event_type", "person", "auction", "bid", "_row_id"] }
```

**Comment by richardchien** [comment]
> The `ToChar(Field(bid, 5:Int32), 'YYYY-MM-DD':Varchar)` is just the column's name!

That will give us a real example of ambiguity mentioned in #4455 😁

**Comment by kwannoel** [comment]
> > @kwannoel do you mean the re-compute of the `TO_CHAR`? I think we can avoid it, but if we can not see it in the flame graph, it can not be the significant issue
> 
> Sorry for my wrong information and misleading. In the plan, only the `StreamProject` before `StreamAppendOnlyHashAgg` does the evaluation of the `TO_CHAR` and the `StreamProject` before `StreamMaterialize` is just get the column from Agg's output. The `ToChar(Field(bid, 5:Int32), 'YYYY-MM-DD':Varchar)` is just the column's name!
> 
> ```
>  StreamMaterialize { columns: [auction, day, total_bids, rank1_bids, rank2_bids, rank3_bids, min_price, max_price, avg_price, sum_price], pk_columns: [auction, day] }
>  └─StreamProject { exprs: [Field(bid, 0:Int32), ToChar(Field(bid, 5:Int32), 'YYYY-MM-DD':Varchar), count, count filter((Field(bid, 2:Int32) < 10000:Int32)), count filter((Field(bid, 2:Int32) >= 10000:Int32) AND (Field(bid, 2:Int32) < 1000000:Int32)), count filter((Field(bid, 2:Int32) >= 1000000:Int32)), min(Field(bid, 2:Int32)), max(Field(bid, 2:Int32)), (sum(Field(bid, 2:Int32)) / count(Field(bid, 2:Int32))), sum(Field(bid, 2:Int32))] }
>    └─StreamAppendOnlyHashAgg { group_key: [Field(bid, 0:Int32), ToChar(Field(bid, 5:Int32), 'YYYY-MM-DD':Varchar)], aggs: [count, count, count filter((Field(bid, 2:Int32) < 10000:Int32)), count filter((Field(bid, 2:Int32) >= 10000:Int32) AND (Field(bid, 2:Int32) < 1000000:Int32)), count filter((Field(bid, 2:Int32) >= 1000000:Int32)), min(Field(bid, 2:Int32)), max(Field(bid, 2:Int32)), sum(Field(bid, 2:Int32)), count(Field(bid, 2:Int32)), sum(Field(bid, 2:Int32))] }
>      └─StreamExchange { dist: HashShard(Field(bid, 0:Int32), ToChar(Field(bid, 5:Int32), 'YYYY-MM-DD':Varchar)) }
>        └─StreamProject { exprs: [Field(bid, 0:Int32), ToChar(Field(bid, 5:Int32), 'YYYY-MM-DD':Varchar), Field(bid, 2:Int32), _row_id] }
>          └─StreamFilter { predicate: (event_type = 2:Int32) }
>            └─StreamRowIdGen { row_id_index: 4 }
>              └─StreamSource { source: "nexmark", columns: ["event_type", "person", "auction", "bid", "_row_id"] }
> ```

Hmm I think `to_char` can still be improved, even if only evaluated in stream project? 🤔 

It is shown that if we get rid of `to_char` it greatly improves performance: https://github.com/risingwavelabs/risingwave/issues/7351#issuecomment-1429474290.

Will investigate against a baseline to see how much performance improvement we can expect (if any).

**Comment by richardchien** [comment]
> > I'm wondering if we reuse the `sum/count` states. 🤔 Will take a look, but seems like no from the rw plan:
> > ```
> > sum(Field(bid, 2:Int32)), count(Field(bid, 2:Int32)), sum(Field(bid, 2:Int32))
> > ```
> > 
> > cc @richardchien
> 
> I think we are not reusing the `sum` when rewriting `avg` to `sum / count`.

After #8200, `sum` is reused🥰

**Comment by fuyufjh** [comment]
Hi, any updates?

**Comment by kwannoel** [comment]
> Hi, any updates?

No updates since @richardchien's improvements. Need to do further investigation on `to_char` and also looking into the recent regression. Seems strange.

My benchmark results from a few weeks ago:

<img width="874" alt="Screenshot 2023-03-13 at 4 28 56 PM" src="https://user-images.githubusercontent.com/47273164/224647086-0685e71a-8cb4-4bf2-b357-9effe645dc40.png">

Recent benchmark results:

<img width="874" alt="Screenshot 2023-03-13 at 4 29 35 PM" src="https://user-images.githubusercontent.com/47273164/224647222-48bbdd3a-52e6-44eb-8bc6-de68ce6efc61.png">

**Comment by kwannoel** [comment]
Should we optimize avg as well? Instead of `avg` -> `sum / count`.

That will save a `project`, and it is what flink does too.

Will try a bench, where `avg` is removed from q17, see if it is worthwhile.

**Comment by st1page** [comment]
> Should we optimize avg as well by writing a specialized agg operator for it? Instead of avg -> sum / count.

how to implement it? it is hard to calculate the AVG incrementally. :thinking:

**Comment by kwannoel** [comment]
> > Should we optimize avg as well by writing a specialized agg operator for it? Instead of avg -> sum / count.
> 
> how to implement it? it is hard to calculate the AVG incrementally. 🤔

I realized it too 🤣 . Is it possible to make the logic of `sum` / `count` internal to the aggregation operator instead?

I'm not sure what is the overhead of project.

**Comment by xxchan** [comment]
> Is it possible to make the logic of sum / count internal to the aggregation operator instead?

Yes, IIRC Flink (Calcite) (&also PG) can do `Calc` in many places. I think the overhead shouldn't be significant..

**Comment by st1page** [comment]
> Is it possible to make the logic of sum / count internal to the aggregation operator instead

a little hard to do it :hot_face: sum and count have the good property that its state is the result.

**Comment by kwannoel** [comment]
Latest query:
```
CREATE SINK nexmark_q17 AS
    SELECT auction,
           to_char(date_time, 'YYYY-MM-DD')                           AS day,
           count(*)                                                   AS total_bids,
           count(*) filter (where price < 10000)                      AS rank1_bids,
           count(*) filter (where price >= 10000 and price < 1000000) AS rank2_bids,
           count(*) filter (where price >= 1000000)                   AS rank3_bids,
           min(price)                                                 AS min_price,
           max(price)                                                 AS max_price,
           avg(price)                                                 AS avg_price,
           sum(price)                                                 AS sum_price
    FROM bid
    GROUP BY auction, to_char(date_time, 'YYYY-MM-DD')
    WITH ( connector = 'blackhole', format = 'append_only', force_append_only = 'true');

```

Latest plan:
```                                                                                                                                      QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 StreamSink { type: append-only, columns: [auction, day, total_bids, rank1_bids, rank2_bids, rank3_bids, min_price, max_price, avg_price, sum_price] }
 └─StreamProject { exprs: [$expr1, $expr2, count, count filter(($expr3 < 10000:Int32)), count filter(($expr3 >= 10000:Int32) AND ($expr3 < 1000000:Int32)), count filter(($expr3 >= 1000000:Int32)), min($expr3), max($expr3), (sum($expr3) / count($expr3)) as $expr4, sum($expr3)] }
   └─StreamAppendOnlyHashAgg { group_key: [$expr1, $expr2], aggs: [count, count filter(($expr3 < 10000:Int32)), count filter(($expr3 >= 10000:Int32) AND ($expr3 < 1000000:Int32)), count filter(($expr3 >= 1000000:Int32)), min($expr3), max($expr3), sum($expr3), count($expr3)] }
     └─StreamExchange { dist: HashShard($expr1, $expr2) }
       └─StreamProject { exprs: [Field(bid, 0:Int32) as $expr1, ToChar(Field(bid, 5:Int32), 'YYYY-MM-DD':Varchar) as $expr2, Field(bid, 2:Int32) as $expr3, _row_id] }
         └─StreamFilter { predicate: (event_type = 2:Int32) }
           └─StreamRowIdGen { row_id_index: 5 }
             └─StreamSource { source: "nexmark", columns: ["event_type", "person", "auction", "bid", "_rw_kafka_timestamp", "_row_id"] }
(8 rows)
```

**Comment by kwannoel** [comment]
> I'm quite curious why there must be a `to_char` in the group key. It seems all the query wants to do is truncate the timestamp into a date. Will we gain better performance if we rewrite it with `date_time::date`? It should even have the same representation in the result. 😄

If throughput shown here https://github.com/risingwavelabs/risingwave/issues/8604#issuecomment-1471871489 is not the result of a bug, we should explore this.

Can nearly 2x throughput, q15 and q16 will benefit significantly too.

**Comment by st1page** [comment]
> Start off with to_char(timestamp, 'YYYY-MM-DD') -> cast to date.
1. Special representation for to_char strings.
2. Can nearly 2x throughput, q15 and q16 will benefit significantly too.

I think it is a kind of optimization that we can pull up the `to_char(timestamp, 'YYYY-MM-DD')` expression to traverse the join or agg. Before it we need to confirm them:
1. ∀t: timestamp, to_char(t, 'YYYY-MM-DD') == to_char(t::date, 'YYYY-MM-DD')
2. ∀t1, t2: timestamp, (t1::date == t2::date)<=> ((to_char(t1, 'YYYY-MM-DD') == to_char(t2, 'YYYY-MM-DD'))

c.c. @chenzl25 @xiangjinwu

**Comment by kwannoel** [comment]
> > Start off with to_char(timestamp, 'YYYY-MM-DD') -> cast to date.
> > 1.Special representation for to_char strings.
> > 2.Can nearly 2x throughput, q15 and q16 will benefit significantly too.
> 
> I think it is a kind of optimization that we can pull up the `to_char(timestamp, 'YYYY-MM-DD')` expression to traverse the join or agg. Before it we need to confirm them:
> 
>     1. ∀t: timestamp, to_char(t, 'YYYY-MM-DD') == to_char(t::date, 'YYYY-MM-DD')
> 
>     2. ∀t1, t2: timestamp, (t1::date == t2::date)<=> ((to_char(t1, 'YYYY-MM-DD') == to_char(t2, 'YYYY-MM-DD'))
> 
> 
> c.c. @chenzl25 @xiangjinwu

Benchmark Details (first column is average throughput, second is max):

https://www.notion.so/risingwave-labs/2043c4721c3d4470b86887a7f7bde062?v=a2068e54f72c46c486a90a8638020524&pvs=4

<img width="1274" alt="Screenshot 2023-03-17 at 1 19 43 PM" src="https://user-images.githubusercontent.com/47273164/225819182-a635739c-bf46-40da-a5ab-4c225288705b.png">


<img width="1274" alt="Screenshot 2023-03-17 at 1 14 56 PM" src="https://user-images.githubusercontent.com/47273164/225818539-9e8f468d-5a23-442d-9c98-477ed4afe7f2.png">

See if I understand correctly:
1. Pull up `to_char` to join or agg.
2. Use t::date before that, which saves the cost (cheaper than manipulating varchars).

**Comment by st1page** [comment]
> See if I understand correctly:
1.Pull up to_char to join or agg.
2.Use t::date before that, which saves the cost (cheaper than manipulating varchars).

Yes!

**Comment by chenzl25** [comment]
Perhaps we could consider benching the following query directly prior to implementing this optimization.
```sql
CREATE SINK nexmark_q17 AS
    SELECT auction,
           to_char(date_time::date, 'YYYY-MM-DD')                           AS day,
           count(*)                                                   AS total_bids,
           count(*) filter (where price < 10000)                      AS rank1_bids,
           count(*) filter (where price >= 10000 and price < 1000000) AS rank2_bids,
           count(*) filter (where price >= 1000000)                   AS rank3_bids,
           min(price)                                                 AS min_price,
           max(price)                                                 AS max_price,
           avg(price)                                                 AS avg_price,
           sum(price)                                                 AS sum_price
    FROM bid
    GROUP BY auction, date_time::date
    WITH ( connector = 'blackhole', format = 'append_only', force_append_only = 'true');
```

**Comment by kwannoel** [comment]
> Perhaps we could consider benching the following query directly prior to implementing this optimization.
> 
> ```sql
> CREATE SINK nexmark_q17 AS
>     SELECT auction,
>            to_char(date_time::date, 'YYYY-MM-DD')                           AS day,
>            count(*)                                                   AS total_bids,
>            count(*) filter (where price < 10000)                      AS rank1_bids,
>            count(*) filter (where price >= 10000 and price < 1000000) AS rank2_bids,
>            count(*) filter (where price >= 1000000)                   AS rank3_bids,
>            min(price)                                                 AS min_price,
>            max(price)                                                 AS max_price,
>            avg(price)                                                 AS avg_price,
>            sum(price)                                                 AS sum_price
>     FROM bid
>     GROUP BY auction, date_time::date
>     WITH ( connector = 'blackhole', format = 'append_only', force_append_only = 'true');
> ```

https://www.notion.so/risingwave-labs/2ce84c09acbb47d39dbd5d14648c575b?pvs=4

<img width="1274" alt="Screenshot 2023-03-17 at 4 49 03 PM" src="https://user-images.githubusercontent.com/47273164/225856841-7f50dcce-ff5a-4cb2-9295-61fbd685aa9a.png">

Throughput looks good.

**Comment by lmatz** [comment]
(380000-300000)/380000 = 21%, which I suppose is all attributed to `to_char` now?

**Comment by kwannoel** [comment]
> (380000-300000)/380000 = 21%, which I suppose is all attributed to `to_char` now?

Well yes, but I'm not sure how much can be attributed to the `evaluation` of `to_char`, especially since it doesn't even show up on flamegraph.

Rather, I think it's more likely due to `to_char` causing date to be represented as a string.

**Comment by kwannoel** [comment]
> > (380000-300000)/380000 = 21%, which I suppose is all attributed to `to_char` now?
> 
> Well yes, but I'm not sure how much can be attributed to the `evaluation` of `to_char`, especially since it doesn't even show up on flamegraph.
> 
> Rather, I think it's more likely due to `to_char` causing date to be represented as a string.
> 
> 

Seems like there's another issue which needs investigating:
How to get to_char to show up on flamegraph. I will making to_char more expensive, in that case it should show up.

If it doesn't, maybe requires some compiler directives to avoid inlining to_char or losing its debug symbols.

**Comment by st1page** [comment]
> Seems like there's another issue which needs investigating: How to get to_char to show up on flamegraph. I will making to_char more expensive, in that case it should show up.
> 
> If it doesn't, maybe requires some compiler directives to avoid inlining to_char or losing its debug symbols.

in 2.14‘s perf.svg
https://file.notion.so/f/s/a1871582-36fd-4c4a-94f7-dc414b476ba4/perf.svg?spaceId=a409f923-73d6-4f09-a08a-7edd6555138b&table=block&id=e31e70e6-0421-46a0-b18a-edfae97785fd&expirationTimestamp=1679137517854&signature=ec1gnh08W7DRBrBP_YSspMLwBIDRoJytSN0MLV1qWc4&downloadName=perf.svg&x=16.7868%25&y=2229

**Comment by st1page** [comment]
I find may be the StreamChunk::compact() cost a lot (https://github.com/singularity-data/risingwave/blob/main/src/stream/src/executor/project.rs). c.c. @kwannoel

**Comment by kwannoel** [comment]
> > Seems like there's another issue which needs investigating: How to get to_char to show up on flamegraph. I will making to_char more expensive, in that case it should show up.
> > If it doesn't, maybe requires some compiler directives to avoid inlining to_char or losing its debug symbols.
> 
> in 2.14‘s perf.svg https://file.notion.so/f/s/a1871582-36fd-4c4a-94f7-dc414b476ba4/perf.svg?spaceId=a409f923-73d6-4f09-a08a-7edd6555138b&table=block&id=e31e70e6-0421-46a0-b18a-edfae97785fd&expirationTimestamp=1679137517854&signature=ec1gnh08W7DRBrBP_YSspMLwBIDRoJytSN0MLV1qWc4&downloadName=perf.svg&x=16.7868%25&y=2229

This is for `StreamChunk::compact()` right, not to_char, just to clarify?

**Comment by kwannoel** [comment]
> I find may be the StreamChunk::compact() cost a lot (https://github.com/singularity-data/risingwave/blob/main/src/stream/src/executor/project.rs#L113). c.c. @kwannoel

<img width="1134" alt="Screenshot 2023-03-17 at 8 08 56 PM" src="https://user-images.githubusercontent.com/47273164/225900846-d1e3701a-3fe2-497e-902c-e76ad5862a3b.png">

Maybe I missed something? Does not seem to be much, just 3.14%.
Given 225,000 r/s, we have 225,000 * (100 * 97) = 231958.

Still worth investigating I guess, I can imagine there are other queries without aggregations where this can be dominant cost.

Also, when is compact called usually? Here are the call sites that show up:

<img width="720" alt="Screenshot 2023-03-17 at 8 18 12 PM" src="https://user-images.githubusercontent.com/47273164/225902196-13a28b1a-b23c-4982-925d-598097016507.png">

**Comment by st1page** [comment]
The graph is the whole StreamProject. and the `to_char` is the `ExprToCharImpl`. What I was saying is that compared with the `to_char` processing, the `compact()` has more cost.
![image](https://user-images.githubusercontent.com/40830455/226111745-b3430763-0cd7-4e01-8b04-26a4c236b7cd.png)

**Comment by lmatz** [comment]
Q0:
```
CREATE SINK nexmark_q0
    AS
    SELECT auction, bidder, price, date_time
    FROM bid
    WITH ( connector = 'blackhole', type = 'append-only');
```

![SCR-20230328-h5u](https://user-images.githubusercontent.com/5791930/228127556-2276459a-aa7b-452c-8bf8-6cab0836616a.png)


Q10:
```
CREATE SINK nexmark_q10 AS
    SELECT auction,
           bidder,
           price,
           date_time,
           TO_CHAR(date_time, 'YYYY-MM-DD') as date,
           TO_CHAR(date_time, 'HH:MI')      as time
    FROM bid
    WITH ( connector = 'blackhole', type = 'append-only');
```
![SCR-20230328-h59](https://user-images.githubusercontent.com/5791930/228127566-8ddedeab-4374-4bc3-8be0-7e88b08c2510.png)

The peak throughputs are 950K/s versus 1M/s.
Two `TO_CHAR` introduced 5% overhead. So not a big issue.

One of the real issues is likely to be: #8802

**Comment by lmatz** [comment]
By looking at the flame graph generated in CI:
https://buildkite.com/risingwavelabs/main-cron/builds/458#0187c454-58bd-439a-bf57-6f774ea41b80

![SCR-20230428-hej](https://user-images.githubusercontent.com/5791930/235054835-3c8c8a56-1645-442b-b814-925cec2b8c90.png)

![SCR-20230428-hen](https://user-images.githubusercontent.com/5791930/235054839-1f5355c8-efef-42f4-9c79-58f8aa1b0c16.png)

Json parser takes 40% more time than Hash Aggregation Executor...
